source/extensions/filters/listener/connect_handler/connect_handler.cc

```diff
@@ -0,0 +1,183 @@
+    nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr,
+};
+
+thread_local uint8_t Filter::buf_[Config::MAX_INSPECT_SIZE];
+
+namespace {
+constexpr absl::string_view response_body = " 200 Connection Established\r\n\r\n";
+}
+
+Network::FilterStatus Filter::onAccept(Network::ListenerFilterCallbacks& cb) {
+  ENVOY_LOG(debug, "connect handler: new connection accepted");
+
+  const Network::ConnectionSocket& socket = cb.socket();
+
+  const absl::string_view transport_protocol = socket.detectedTransportProtocol();
+  if (!transport_protocol.empty() && transport_protocol != "raw_buffer") {
+    ENVOY_LOG(trace, "connect handler: cannot inspect http protocol with transport socket {}",
+              transport_protocol);
+    return Network::FilterStatus::Continue;
+  }
+
+  cb_ = &cb;
+  const ParseState parse_state = onRead();
+  switch (parse_state) {
+  case ParseState::Error:
+    // As per discussion in https://github.com/envoyproxy/envoy/issues/7864
+    // we don't add new enum in FilterStatus so we have to signal the caller
+    // the new condition.
+    cb.socket().close();
+    return Network::FilterStatus::StopIteration;
+  case ParseState::Done:
+    return Network::FilterStatus::Continue;
+  case ParseState::Continue:
+    // do nothing but create the event
+    cb.socket().ioHandle().initializeFileEvent(
+        cb.dispatcher(),
+        [this](uint32_t events) {
+          ASSERT(events == Event::FileReadyType::Read);
+          ENVOY_LOG(trace, "connect handler event: {}", events);
+
+          const ParseState parse_state = onRead();
+          switch (parse_state) {
+          case ParseState::Error:
+            cb_->socket().ioHandle().resetFileEvents();
+            cb_->continueFilterChain(false);
+            break;
+          case ParseState::Done:
+            cb_->socket().ioHandle().resetFileEvents();
+            // Do not skip following listener filters.
+            cb_->continueFilterChain(true);
+            break;
+          case ParseState::Continue:
+            // do nothing but wait for the next event
+            break;
+          }
+        },
+        Event::PlatformDefaultTriggerType, Event::FileReadyType::Read);
+    return Network::FilterStatus::StopIteration;
+  }
+  ENVOY_BUG(false, "Unexpected ParseState value for connect handler");
+}
+
+ParseState Filter::onRead() {
+  auto result = cb_->socket().ioHandle().recv(buf_, Config::MAX_INSPECT_SIZE, MSG_PEEK);
+  ENVOY_LOG(trace, "connnect handler: recv: {}", result.return_value_);
+  if (!result.ok()) {
+    if (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again) {
+      return ParseState::Continue;
+    }
+    config_->stats().read_error_.inc();
+    return ParseState::Error;
+  }
+
+  // Remote closed
+  if (result.return_value_ == 0) {
+    return ParseState::Error;
+  }
+  return parseConnect(absl::string_view(reinterpret_cast<const char*>(buf_), result.return_value_));
```

Comment:
No need to instantiate a temp string_view here.

Reply:
done